File: src/cf_ip_blacklist/analytics.py
from __future__ import annotations

import re
from datetime import datetime
from hashlib import sha256
from typing import Any
from urllib.parse import urlsplit

import httpx

from .errors import CloudflareError
from .http import request_with_retry
from .models import Observation
from .policy import normalize_ip

QUERY = """
query Requests($zone: String!, $start: DateTime!, $end: DateTime!) {
  viewer { zones(filter: {zoneTag: $zone}) {
    series: httpRequestsAdaptiveGroups(
      limit: 1000, filter: {datetime_geq: $start, datetime_lt: $end},
      orderBy: [count_DESC]
    ) {
      dimensions { clientIP edgeResponseStatus clientRequestPath }
      avg { sampleInterval }
      count
    }
  } }
}
"""


def _error_message(payload: dict[str, Any]) -> str:
    errors = payload.get("errors") or payload.get("messages") or []
    return "; ".join(str(item.get("message", item)) for item in errors)


def _path_without_query(value: str) -> str:
    return urlsplit(value).path


def _is_suspicious_path(path: str, patterns: tuple[str, ...]) -> bool:
    return any(re.search(pattern, path, re.IGNORECASE) for pattern in patterns)


def parse_grouped(
    payload: dict[str, Any],
    zone_id: str,
    observed_at: datetime,
    suspicious_path_patterns: tuple[str, ...] = (),
) -> list[Observation]:
    if payload.get("errors"):
        raise CloudflareError(f"GraphQL error: {_error_message(payload)}")
    try:
        rows = payload["data"]["viewer"]["zones"][0]["series"]
    except (KeyError, IndexError, TypeError) as exc:
        raise CloudflareError("GraphQL response missing required series fields") from exc
    observations: list[Observation] = []
    for row in rows:
        try:
            dimensions = row["dimensions"]
            ip = normalize_ip(dimensions["clientIP"])
            response_status = int(dimensions["edgeResponseStatus"])
            count = int(row["count"])
            path = _path_without_query(str(dimensions["clientRequestPath"]))
            sample_interval = int(row.get("avg", {}).get("sampleInterval", 1) or 1)
            if sample_interval < 1:
                raise ValueError("invalid sample interval")
        except (KeyError, TypeError, ValueError) as exc:
            raise CloudflareError("invalid grouped analytics row") from exc
        fingerprint = sha256(
            f"{zone_id}:{ip}:{path}:{response_status}:{observed_at.isoformat()}".encode()
        ).hexdigest()[:16]
        observations.append(
            Observation(
                ip=ip,
                zone_id=zone_id,
                observed_at=observed_at,
                observed_requests=count,
                weighted_requests=float(count * sample_interval),
                paths=[path],
                suspicious_paths=int(_is_suspicious_path(path, suspicious_path_patterns)),
                error_requests=count if response_status >= 400 else 0,
                sampled=sample_interval > 1,
                sample_interval=sample_interval if sample_interval > 1 else None,
                fingerprint=fingerprint,
            )
        )
    return observations


class AnalyticsClient:
    def __init__(
        self,
        client: httpx.Client,
        url: str,
        max_retries: int = 3,
        suspicious_path_patterns: tuple[str, ...] = (),
    ) -> None:
        self.client = client
        self.url = url
        self.max_retries = max_retries
        self.suspicious_path_patterns = suspicious_path_patterns

    def collect(self, zone_id: str, start: datetime, end: datetime) -> list[Observation]:
        response = request_with_retry(
            self.client,
            "POST",
            self.url,
            self.max_retries,
            json={
                "query": QUERY,
                "variables": {
                    "zone": zone_id,
                    "start": start.isoformat(),
                    "end": end.isoformat(),
                },
            },
        )
        if response.status_code >= 400:
            raise CloudflareError(f"GraphQL HTTP {response.status_code}")
        return parse_grouped(response.json(), zone_id, end, self.suspicious_path_patterns)

File: src/analytics.rs
use crate::{
    error::{BlockholeError, Result},
    http::request,
    models::{Observation, Subject},
};
use chrono::{DateTime, Utc};
use regex::RegexSet;
use reqwest::blocking::Client;
use serde::Deserialize;
use sha2::{Digest, Sha256};
use url::Url;
const QUERY: &str = "query Requests($zone: String!, $start: DateTime!, $end: DateTime!) { viewer { zones(filter: {zoneTag: $zone}) { series: httpRequestsAdaptiveGroups(limit: 1000, filter: {datetime_geq: $start, datetime_lt: $end}, orderBy: [count_DESC]) { dimensions { clientIP edgeResponseStatus clientRequestPath } avg { sampleInterval } count } } } }";
#[derive(Deserialize)]
struct Payload {
    data: Option<Data>,
    errors: Option<Vec<ApiError>>,
}
#[derive(Deserialize)]
struct ApiError {
    message: String,
}
#[derive(Deserialize)]
struct Data {
    viewer: Viewer,
}
#[derive(Deserialize)]
struct Viewer {
    zones: Vec<Zone>,
}
#[derive(Deserialize)]
struct Zone {
    series: Vec<Row>,
}
#[derive(Deserialize)]
struct Row {
    dimensions: Dimensions,
    avg: Option<Average>,
    count: u64,
}
#[derive(Deserialize)]
struct Dimensions {
    #[serde(rename = "clientIP")]
    client_ip: String,
    #[serde(rename = "edgeResponseStatus")]
    status: u16,
    #[serde(rename = "clientRequestPath")]
    path: String,
}
#[derive(Deserialize)]
struct Average {
    #[serde(rename = "sampleInterval")]
    sample_interval: Option<f64>,
}
pub fn parse(
    payload: &str,
    zone_id: &str,
    observed_at: DateTime<Utc>,
    pattern_set: &RegexSet,
) -> Result<Vec<Observation>> {
    let payload: Payload = serde_json::from_str(payload)?;
    if let Some(errors) = payload.errors {
        return Err(BlockholeError::Cloudflare(
            errors
                .into_iter()
                .map(|e| e.message)
                .collect::<Vec<_>>()
                .join("; "),
        ));
    }
    let rows = payload
        .data
        .ok_or_else(|| BlockholeError::Cloudflare("GraphQL response missing data".into()))?
        .viewer
        .zones
        .into_iter()
        .next()
        .ok_or_else(|| BlockholeError::Cloudflare("GraphQL response missing zone".into()))?
        .series;
    rows.into_iter()
        .map(|row| {
            let ip = Subject::parse(&row.dimensions.client_ip)?;
            let path = Url::parse(&format!(
                "https://placeholder.invalid{}",
                row.dimensions.path
            ))
            .map_err(|e| BlockholeError::Cloudflare(e.to_string()))?
            .path()
            .to_string();
            let interval = row.avg.and_then(|a| a.sample_interval).unwrap_or(1.0);
            if !interval.is_finite() || interval <= 0.0 {
                return Err(BlockholeError::Cloudflare("invalid sample interval".into()));
            }
            let suspicious = pattern_set.is_match(&path);
            let mut hasher = Sha256::new();
            hasher.update(
                format!(
                    "{zone_id}:{ip}:{path}:{}:{observed_at}",
                    row.dimensions.status
                )
                .as_bytes(),
            );
            let fingerprint = format!("{:x}", hasher.finalize())[..16].to_string();
            Ok(Observation {
                ip,
                zone_id: zone_id.into(),
                observed_at,
                observed_requests: row.count,
                weighted_requests: row.count as f64 * interval,
                paths: vec![path],
                suspicious_paths: u64::from(suspicious),
                error_requests: if row.dimensions.status >= 400 {
                    row.count
                } else {
                    0
                },
                sampled: interval > 1.0,
                sample_interval: (interval > 1.0).then_some(interval),
                fingerprint,
            })
        })
        .collect()
}
pub fn collect(
    client: &Client,
    url: &str,
    retries: usize,
    zone: &str,
    start: DateTime<Utc>,
    end: DateTime<Utc>,
    pattern_set: &RegexSet,
) -> Result<Vec<Observation>> {
    let body = serde_json::json!({"query": QUERY, "variables": {"zone": zone, "start": start.to_rfc3339(), "end": end.to_rfc3339()}});
    let response = request(client, reqwest::Method::POST, url, retries, Some(body))?;
    if response.status().is_client_error() || response.status().is_server_error() {
        return Err(BlockholeError::Cloudflare(format!(
            "GraphQL HTTP {}",
            response.status()
        )));
    }
    parse(&response.text()?, zone, end, pattern_set)
}
